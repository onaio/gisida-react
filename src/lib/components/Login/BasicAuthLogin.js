import React, { Component } from 'react';
import Cookie from 'js-cookie';

export const isLoggedIn = function () {
    return Cookie.get('dsauth') === "true";
};


class BasicAuthLogin extends Component {
    constructor(props) {
        super(props);
        this.state = { loginError: false };
    }

    handleLogin(password) {
        const { appPassword } = this.props;
        if (appPassword.includes(password)) {
            Cookie.set('dsauth', true);
            location.reload();
        } else {
            this.setState({ loginError: true });
        }
    }

    render() {
        const { loginImageStyle } = this.props;

        return (
            <div className="login">
                <div>
                    <form className="login-form" onSubmit={(e) => { e.preventDefault(); this.handleLogin(this.password.value); }}>
                        <div className="form-group">
                            <div className="brand-login" style={loginImageStyle}></div>
                            {this.state.loginError === true ? <div className="alert alert-danger">Incorrect password.</div> : null}
                            <label htmlFor="password">Enter your Password</label>
                            <input className="form-control" type="password" ref={(input) => { this.password = input; }} autoFocus />
                        </div>
                        <button className="btn btn-default center-block" type="submit">Log In</button>
                    </form>
                </div>
            </div>);
    }
}

export default BasicAuthLogin