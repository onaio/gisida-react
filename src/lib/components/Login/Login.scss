.login {
    width: 100%;
    background: #fff;
    color: #666;
    height: 100%;
    position: fixed;
    .brand-login {
        display: block;
        width: 200px;
        height: 200px;
        text-align: center;
        color: #ccc;
        background-size: 70% !important;
        text-indent: -9999px;
        margin: 10% auto 0;
    }
    .login-form {
        width: 200px;
        margin: 0 auto;
        .login input {
            background: rgba(0, 0, 0, 0.8);
        }
    }
}