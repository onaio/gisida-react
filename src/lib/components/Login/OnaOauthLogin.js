import React, { Component } from 'react';
import { useOAuthLogin } from '@onaio/gatekeeper/dist/components/login';
import { AuthorizationGrantType } from '@onaio/gatekeeper';

class OnaOauthLogin extends Component {
  constructor(props) {
    super(props);
    this.state = {
      authorizationUris: {},
    };
  }

  componentDidMount() {
    const clientID = process.env.REACT_APP_GISIDA_CANOPY_CLIENT_ID
    const redirectUri = location.protocol + '//' + location.host + '/callback';
    const providers = {
      onadata: {
        accessTokenUri: '',
        authorizationUri: 'https://api.ona.io/o/authorize/',
        clientId: clientID,
        redirectUri: redirectUri,
        scopes: ['read', 'write'],
        state: 'abc',
        userUri: 'https://api.ona.io/user',
      },
    };
    const options = {
      providers,
      authorizationGrantType: AuthorizationGrantType.IMPLICIT,
    };
    const authorizationUris = useOAuthLogin(options);
    this.setState({
      authorizationUris,
    });
  }

  getProviderKey() {
    return Object.keys(this.state.authorizationUris).filter(key => key === 'onadata')[0];
  }

  render() {
    const { publicPassword, publicUsername } = this.props;
    const providerKey = this.getProviderKey();

    return (
      <form className="login-form">
        <div className="form-group">
          {publicPassword && publicUsername ? (
            <div>
              <small>Username: {publicUsername}</small>
              <br />
              <small>Password: {publicPassword}</small>
            </div>
          ) : null}
        </div>
        <a
          className="btn btn-default center-block btn-block"
          href={this.state.authorizationUris[providerKey]}
        >
          Login
            </a>
      </form>
    );
  }
}

export default OnaOauthLogin
