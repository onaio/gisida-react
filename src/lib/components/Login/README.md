# Login Component

The component returns a `BasicAuthLogin` component for Basic Authentication or a `OnaOauthLogin` component for
Ona OAuth2 Implicit Grant Type Authentication based on configuration.


## Basic Authentication (Not recommended)

Caution: basic authentication is not a secure method of protecting sensitive data, it should only be used to grant basic access to view sites which do not include sensitive information

Update your client project with the following changes:

In `site-config.json` file, add the `password` property with a value for
the correct password to be checked against.

```js
{
    ...
    APP: {
        ...
        "password": ["password123"],
        ...
    }
    ...
}
```

In `index.js` file

```js

import { Login, isBasicAuthLoggedIn } from 'gisida-react'
....

if (isBasicAuthLoggedIn()) {
  ReactDOM.render(
    <Provider store={store}>
        <App>
            <TitleBar />
            ...
        </App>
    </Provider>,
    rootElement
  );
} else {
  ReactDOM.render(
    <Provider store={store}>
      <Login />
    </Provider>,
    rootElement
  );
}
```

## Ona OAuth2 Implicit Grant Type Authentication (Recommended)

### Prerequisites

You must have an Ona application client ID. If, you do not have the client ID, [Register your client application with Ona](https://api.ona.io/static/docs/authentication.html#using-oauth2-with-the-ona-api) to obtain yout client ID


### Set Up

Update your client project with the following changes:

[Install `dotenv-webpack`](https://www.npmjs.com/package/dotenv-webpack) and [update the client `webpack.config.js`](https://www.npmjs.com/package/dotenv-webpack#add-it-to-your-webpack-config-file) to use `dotenv-webpack`

Add `.env` to `.gitignore`

Add a `.env` file to the root directory of your client project and assign the client ID to the key
`REACT_APP_GISIDA_CANOPY_CLIENT_ID`

```
REACT_APP_GISIDA_CANOPY_CLIENT_ID=<Your client Id>
```

Add a `.env.sample` a best practice that is used as an information file for all team members to know what keys and values may be needed. Add the variable

```
REACT_APP_GISIDA_CANOPY_CLIENT_ID=<Your client Id>
```

In `site-config.json` file, remove the property password if it exists.

```js
{
    ...
    APP: {
        ...
        "password": ["password123"], # Remove this
        ...
    }
    ...
}
```

In `index.js` file

```js

import { Login, Callback, Router } from 'gisida-react'
import { SupAuth } from 'gisidia'
....

const AppView = (
    <App>
        <TitleBar />
        ...
    </App>
);
const oauthclientID = process.env.REACT_APP_GISIDA_CANOPY_CLIENT_ID;
const LoginView = <Login oauthclientID={oauthclientID}/>;
const CallbackView = <Callback />;

ReactDOM.render(
  <Provider store={store}>
    <Router.Wrapper>
      <Router.PrivateRoute
        exact
        path="/"
        auth={SupAuth.defaultSupViewAuthC}
        component={() => AppView}
      />
      <Router.PublicRoute path="/login" component={() => LoginView} />
      <Router.PublicRoute path="/callback" component={() => CallbackView} />
    </Router.Wrapper>
  </Provider>,
  rootElement
);
```

### Deployment

Export your client ID as an environmental variable in your production server and restart. There are many
ways to achieving this and you are not confined to using the one below.

```sh
export REACT_APP_GISIDA_CANOPY_CLIENT_ID=<Your client id>
```

It is also recommended to use an additional layer of server side protection to secure static resources from
being accessible without authentication.