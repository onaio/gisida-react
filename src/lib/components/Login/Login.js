import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Cookie from 'js-cookie';
import { SupAuth, isTokenExpired } from 'gisida';
import './Login.scss';
import BasicAuthLogin from './BasicAuthLogin/BasicAuthLogin'
import OnaOauthLogin from './OnaOauthLogin/OnaOauthLogin'

const { defaultUnSupAuthZ: deAuthZ } = SupAuth;

const mapStateToProps = (state, ownProps) => {
    const { APP } = state;
    return {
        appIcon: APP.appIcon,
        loginIcon: APP.appLoginIcon,
        appPassword: APP.password,
        appNameDesc: APP.appNameDesc,
        oauthProvider: ownProps.oauthProvider || 'onadata'
    };
};

export const killSession = () => {
    Cookie.remove('dsauth');
    localStorage.removeItem('expiry_time');
    deAuthZ();
}
  
export const isLoggedIn = function() {
    const hasCookie = Cookie.get('dsauth') === "true";
    const tokenIsExpired = isTokenExpired();
    if (!hasCookie || tokenIsExpired) {
        killSession();
        return false;
    }
    return true;
};

class Login extends Component {
    render() {
        if (!this.props.appPassword && !this.props.oauthclientID) {
            return null;
        }

        const { appPassword, appNameDesc, loginIcon, appIcon,
            oauthclientID, oauthProvider, publicPassword, publicUsername } = this.props;

        return (
            <div className="login">
                <img className="brand-login" src={loginIcon || appIcon} title={appNameDesc} alt={appNameDesc}></img>
                {appPassword ?
                    <BasicAuthLogin appPassword={appPassword} />
                    :
                    <OnaOauthLogin clientID={oauthclientID} provider={oauthProvider} publicPassword= {publicPassword} publicUsername = {publicUsername}/>
                }
            </div>
        )
    }
}

Login.propTypes = {
    clientID: PropTypes.string,
    appIcon: PropTypes.string,
    loginIcon: PropTypes.string,
    appPassword: PropTypes.arrayOf(PropTypes.string),
    appNameDesc: PropTypes.string
}

export default connect(mapStateToProps)(Login);
