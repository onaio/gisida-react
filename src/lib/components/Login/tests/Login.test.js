import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import Login from './../Login';
import configureMockStore from "redux-mock-store";

const mockStore = configureMockStore();
const initStore = {
    APP: {
        appIcon: '/path/to/appicon.jpg',
        appLoginIcon: '/path/to/loginicon.jpg',
        appNameDesc: 'App name description',
        password: ['$0me $s+rong p@$$w0rd']
    }
}
const store = mockStore(initStore);

describe('/src/lib/components/Login', () => {
    it('renders correctly', () => {
        const wrapper = mount(
            <Provider store={store}>
                <Login />
            </Provider>
        )
        expect(wrapper.find('div.login').length).toBe(1)
        expect(wrapper.find('img').props()).toMatchSnapshot('img')
        wrapper.unmount()
    })

    it('renders app icon as image if login icon is not provided', () => {
        const storeNoLoginIcon = mockStore({
            ...initStore,
            APP: {
                ...initStore.APP,
                appLoginIcon: undefined
            }
        });
        const wrapper = mount(
            <Provider store={storeNoLoginIcon}>
                <Login />
            </Provider>
        )
        expect(wrapper.find('img').prop('src')).toEqual('/path/to/appicon.jpg')
        wrapper.unmount()
    })

    it('renders null if basic auth password and ona oauth client ID is missing', () => {
        const storeNull = mockStore({ APP: {} });
        const wrapper = mount(
            <Provider store={storeNull}>
                <Login />
            </Provider>
        )
        expect(wrapper.isEmptyRender()).toBe(true)
        wrapper.mount()
    });

    it('renders basic auth correctly', () => {
        const wrapper = mount(
            <Provider store={store}>
                <Login />
            </Provider>
        )
        expect(wrapper.find('BasicAuthLogin').props()).toEqual({
            appPassword: ['$0me $s+rong p@$$w0rd']
        })
        wrapper.mount()
    })

    it('renders ona oauth correctly', () => {
        const storeOnaOauth = mockStore({
            ...initStore,
            APP: {
                ...initStore.APP,
                password: undefined,

            },
        });
        const props = {
            clientID: 'client_id'
        }
        const wrapper = mount(
            <Provider store={storeOnaOauth}>
                <Login {...props} />
            </Provider>
        )
        expect(wrapper.find('OnaOauthLogin')).toBeTruthy()
        wrapper.mount()
    })
})
