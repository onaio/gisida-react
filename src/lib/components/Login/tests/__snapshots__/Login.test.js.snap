// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/src/lib/components/Login renders correctly: img 1`] = `
Object {
  "alt": "App name description",
  "className": "brand-login",
  "src": "/path/to/loginicon.jpg",
  "title": "App name description",
}
`;
